src/Microsoft.Azure.SignalR/EndpointProvider/ServiceEndpointManager.cs

```diff
@@ -30,5 +44,74 @@ public override IServiceEndpointProvider GetEndpointProvider(ServiceEndpoint end
 
             return new ServiceEndpointProvider(endpoint, _options);
         }
+
+        private void OnChange(ServiceOptions options)
+        {
+            Log.DetectConfigurationChanges(_logger);
+
+            // Reset local cache and validate result
+            SetValuableEndpoints(GetEndpoints(options));
+            if (Endpoints.Length == 0)
+            {
+                throw new ArgumentException(ServiceEndpointProvider.ConnectionStringNotFound);
+            }
+            var updatedEndpoints = Endpoints.ToList();
+
+            var endpoints = GetChangedEndpoints(updatedEndpoints);
+
+            // RenameServiceEndpoints(endpoints.RenamedEndpoints);
+
+            _ = DoScaleAsync(endpoints.AddedEndpoints, endpoints.RemovedEndpoints);
+
+            _endpointsStore = updatedEndpoints.ToList();
+        }
+
+        private Task DoScaleAsync(IReadOnlyList<ServiceEndpoint> addedEndpoints, IReadOnlyList<ServiceEndpoint> removedEndpoints)
+        {
+            // First add then remove to minor the affect to new clients.
+            // Update EndpointType follow same process.
+            // AddServiceEndpoints(addedEndpoints);
+            // 
+            // RemoveServiceEndpoint(removedEndpoints);
+
+            return Task.CompletedTask;
+        }
+
+        private (IReadOnlyList<ServiceEndpoint> AddedEndpoints, 
+            IReadOnlyList<ServiceEndpoint> RemovedEndpoints,
+            IReadOnlyList<ServiceEndpoint> RenamedEndpoints)
+            GetChangedEndpoints(IEnumerable<ServiceEndpoint> updatedEndpoints)
+        {
+            var addedEndpoints = updatedEndpoints.Except(_endpointsStore, new ServiceEndpointWeakComparer()).ToList();
+            var removedEndpoints = _endpointsStore.Except(updatedEndpoints, new ServiceEndpointWeakComparer()).ToList();
+
+            var renamedEndpoints = updatedEndpoints.Except(_endpointsStore).Except(addedEndpoints).ToList();
+
+            return (AddedEndpoints: addedEndpoints, RemovedEndpoints: removedEndpoints, RenamedEndpoints: renamedEndpoints);
+        }
+
+        private sealed class ServiceEndpointWeakComparer : IEqualityComparer<ServiceEndpoint>
+        {
+            public bool Equals(ServiceEndpoint x, ServiceEndpoint y)
+            {
+                return x.ConnectionString == y.ConnectionString && x.EndpointType == y.EndpointType;
```

Comment:
>ConnectionString [](start = 25, length = 16)

ConnectionString contains the access key? are the 2 ServiceEndpoints considered as the same one if one is using primary access key and another is using secondary access key?

Reply:
Right. Previously thought about regenerated keys and just realized no need to reconnect even in this case, so Endpoint compare should be enough. Will update.

---
In reply to: [377514199](https://github.com/Azure/azure-signalr/pull/801#discussion_r377514199) [](ancestors = 377514199)